notes/conds.md

```diff
@@ -0,0 +1,84 @@
+
+| Operator     | Function Signature                | Description                                | Example                                        |
+| ------------ | --------------------------------- | ------------------------------------------ | ---------------------------------------------- |
+| `/${string}` |`[] -> Any`                        | Selects value using JSON Pointer            | `["/foo/0"]`                                  |
+|              |`[] -> Any`                        | Selects whole value like "" in JSON Pointer | `[]`                                          |
+| `==`         |`[Exp<Any>, Exp<Any>] -> Bool`     | Equality                                    | `["==", ["/size"], 1]`                        |
+| `<`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than                                   | `["<",  ["/size"], 1024]`                     |
+| `<=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than or equal                          | `["<=", ["/size"], 1024]`                     |
+| `>`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than                                | `[">", ["@", "/size"], 0]`                    |
+| `>=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than or equal                       | `[">=", ["@", "/size"], 32]`                  |
+| `not`        |`[Expr<Bool>] -> Bool`             | Negation                                    | `["not", ["<=", ["/size"], 128]`              |
+| `some`       |`[Expr<Any[]>, Expr<Bool>] -> Bool`| Collection predicate                        | `["some", ["/to"], ["==", [], "hi@web.mail"]]`|
+| `match`      |`[Expr<Str>, Exp<Str>] -> Bool`    | String wildcard                             | `["match", ["/to"], "*@foo.com"]`             |
+| `or`         |`Exp<Bool>[] -> Bool`              | Logical **or** combinator                   | `["or", [">", ["/n"], 0], ["==", ["/n"], 0]]` |
+| `and`        |`Exp<Bool>[] -> Bool`              | Logical **and** combinator                  | `["and", [">", ["/n"], 0], ["<", ["/n"], 10]]`|
+
+## Examples
+
+Delegates `email/send` command on `did:key:alice` with a following constraint: Email is not send to address on `gmail.com`, or someone from `fission.codes` should be cc-ed.
+
+```js
+{
+  "sub": "did:key:alice",
+  "iss": "did:key:alice",
+  "aud": "did:key:bob",
+  "cmd": "email/send",
+  "cond": [
+    ["or",
+      // no recipient should be at `@gmail.com`
+      ["every", ["/to"], ["not", ["match", [], "*@gmail.com"]]],
+      // or cc should include @fission.codes
+      ["some", ["/cc"], ["match", [], "*@fission.codes"]]
+    ]
+  ]
+}
+```
+
+## logic programming language
+
+We are also considering logic programming alternative inspired by datomic query styntax.
+
+### concrete proposal for the inital release
+
+Here's a sketch that shows the features
+
+```js
+{
+  "iss": "did:key:alice",
+  // ...
+  "policy": [
+    ["var", ["?x", "?y"], // The `?` here is just by convention
```

Comment:
I don't feel _deeply_ about this, and actually I agree with you in a language like EDN, but we don't have control over the syntax like they do. I think this forces us closer to miniKanren, which has `fresh` (I just renamed `fresh` to `var` here)

Reply:
The other deisgn that I played with was pulling selectors into their own top-level field `{iss: ..., select: {...}, policy: [...]}`, kind of like how SQL does it, but I kind of like the together.

Reply:
> I don't feel _deeply_ about this, and actually I agree with you in a language like EDN, but we don't have control over the syntax like they do. I think this forces us closer to miniKanren, which has `fresh` (I just renamed `fresh` to `var` here)

I'm not sure I understand rational here. Sure in EDN they have symbols and in JSON we do not. That said we could introduce notation that can be interpreted as symbols and get the same effect. Which is kind of what DAG-JSON already does for links and bytes.

What am I missing ?

Reply:
> I don't feel deeply about this

In that case I would suggest to leave it out for now and we can introduce in the future if we find necessary, datomic also has it BTW but I don't think it's common to use them.

Reply:
> ```js
> {
>   iss: "did:...",
>   select: {
>     "x": ["args", "some.path"],
>     "y": ["x", "more.path"]
>   },
>   policy: [...]
> }
> ```

It would be even simpler as follows

```js
{
  iss: "did:...",
  select: {
    "x": "some.path",
    "y": "x.more.path"
  },
  policy: [...]
}
```

But I also don't find it to be better personally.

Reply:
I made a change to be the smallest possible version of this: just "treat these as variables below", no selectors

Reply:
Just saw this:

> I'm not sure I understand rational here. Sure in EDN they have symbols and in JSON we do not. That said we could introduce notation that can be interpreted as symbols and get the same effect. Which is kind of what DAG-JSON already does for links and bytes.
>
> What am I missing ?

Yeah, even in IPLD there's the _possibility_ of namespace conflict. I was looking for a way to seperate these out, but your suggestion of escaping is a good one. "?foo? for variables, and `\?foo` for the string `"?foo"` 👍

Reply:
Removed the `vars` field 👍